AOC [23.09](https://adventofcode.com/2023/day/9) Mirage Maintenance

Input is data from an environmental sensor. Each line contains the history of a single variable.  Figure out an equation to predict next value for each line.

Procedure: 
1. for each sequence of numbers, create a new sequence that is the difference between the numbers
* 1, 3, 6, 10, 15, 21 --> 2, 3, 4, 5, 6
1. Keep doing this until all the differences are zeros:
* 1, 3, 6, 10, 15, 21 --> 2, 3, 4, 5, 6 --> 1, 1, 1 --> 0, 0, 0
1. start with zero and add the last digit in each line:
* 0 + 1 = 1
* 1 + 6 = 7
* 7 + 21 = 28

Find the sum of all the lists in the input.

Solution:
1. write a function that takes a list of numbers and returns the next value in the list
1. the function should do the process described above (create a series of lists)


In [17]:
test_data = '''0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45'''

data = [line for line in test_data.split('\n')]
data = [list(map(int, x.split())) for x in data]

data

[[0, 3, 6, 9, 12, 15], [1, 3, 6, 10, 15, 21], [10, 13, 16, 21, 30, 45]]

In [18]:
[list(map(int, x.split())) for x in line for line in test_data.split('\n')]

NameError: name 'line' is not defined

In [22]:
def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i+1] - sequence[i])
    return diffs

next_value = 0
for sequence in data:
    next_row = find_diffs(sequence)
    list_of_sequences.append(next_row)
    if sum(next_row) == 0:
        next_val = find_next_val(


SyntaxError: incomplete input (2662722400.py, line 12)

In [36]:
def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i+1] - sequence[i])
    return diffs


last_val = 10**6
curr_seq = [10 , 13  ,16 , 21 , 30,  45 ]
next_val = curr_seq[-1] # get the last value in the curr sequence
print(curr_seq,last_val, next_val)

while last_val > 0:
    diffs = find_diffs(curr_seq) # find the last val in the list of differences
    last_val = diffs[-1]
    next_val += last_val
    curr_seq = diffs
    print(curr_seq,last_val, next_val)



[10, 13, 16, 21, 30, 45] 1000000 45
[3, 3, 5, 9, 15] 15 60
[0, 2, 4, 6] 6 66
[2, 2, 2] 2 68
[0, 0] 0 68


In [38]:

data = [line for line in test_data.split('\n')]
data = [list(map(int, x.split())) for x in data]

data

def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i+1] - sequence[i])
    return diffs

tally = 0
for curr_seq in data:
    last_val = 10**6
    next_val = curr_seq[-1]
    print(curr_seq)
    while last_val > 0:
        diffs = find_diffs(curr_seq) # find the last val in the list of differences
        last_val = diffs[-1]
        next_val += last_val
        curr_seq = diffs
    print(f"next element = {next_val}")
    tally += next_val
print(tally)
        


[0, 3, 6, 9, 12, 15]
next element = 18
[1, 3, 6, 10, 15, 21]
next element = 28
[10, 13, 16, 21, 30, 45]
next element = 68
114


Test data works fine. Let's try production data...

In [56]:
with open ('2309input.txt') as f_in:
    data = [line for line in f_in.read().split('\n')]

#data = [line for line in test_data.split('\n')]
data = [list(map(int, x.split())) for x in data]

data

def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i+1] - sequence[i])
    return diffs

tally = 0
for curr_seq in data:
    last_val = 10**10
    next_val = curr_seq[-1]
    #print(curr_seq)
    while sum(curr_seq) != 0:
        diffs = find_diffs(curr_seq) # find the last val in the list of differences
        last_val = diffs[-1]
        next_val += last_val
        curr_seq = diffs
    #print(f"next element = {next_val}")
    tally += next_val

print(f"{tally:,},  {tally}")


1,877,825,182,  1877825182


Run #1:
* 1,879,448,560 is too high a value.

Run #2 - change the test from
`last_val > 0` to `last_val != 0`
* 1,877,825,200 is still too high

Run #3 - change the test to: `while sum(curr_seq) != 0`
* 1,877,825,182 is too *low*


In [57]:
last_val = 10**6
curr_seq = data[3]
next_val = curr_seq[-1] # get the last value in the curr sequence
print(curr_seq,last_val, next_val)

while sum(curr_seq) != 0:
    diffs = find_diffs(curr_seq) # find the last val in the list of differences
    last_val = diffs[-1]
    next_val += last_val
    curr_seq = diffs
    print(curr_seq,last_val, next_val)

[14, 21, 42, 87, 173, 334, 631, 1162, 2072, 3563, 5904, 9441, 14607, 21932, 32053, 45724, 63826, 87377, 117542, 155643, 203169] 1000000 203169
[7, 21, 45, 86, 161, 297, 531, 910, 1491, 2341, 3537, 5166, 7325, 10121, 13671, 18102, 23551, 30165, 38101, 47526] 47526 250695
[14, 24, 41, 75, 136, 234, 379, 581, 850, 1196, 1629, 2159, 2796, 3550, 4431, 5449, 6614, 7936, 9425] 9425 260120
[10, 17, 34, 61, 98, 145, 202, 269, 346, 433, 530, 637, 754, 881, 1018, 1165, 1322, 1489] 1489 261609
[7, 17, 27, 37, 47, 57, 67, 77, 87, 97, 107, 117, 127, 137, 147, 157, 167] 167 261776
[10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10] 10 261786
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] 0 261786
